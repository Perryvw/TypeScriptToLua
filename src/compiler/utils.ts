import { create as createResolve } from "enhanced-resolve";
import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import { assert } from "../utils";
import * as diagnosticFactories from "./diagnostics";

// https://github.com/webpack/enhanced-resolve/blob/0001f80dacf033ac4a0e690b2766e0965c458266/lib/Resolver.js#L280-L288
export const isResolveError = (error: unknown): error is Error & { details: string } =>
    error instanceof Error && "details" in error;

const resolveConfigImport = createResolve.sync({ extensions: [".js", ".ts", ".tsx"] });

export function loadConfigImport(
    kind: string,
    optionName: string,
    basedir: string,
    query: string,
    importName = "default"
): { error?: ts.Diagnostic; result?: unknown } {
    if (typeof query !== "string") {
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "string") };
    }

    let resolved: string;
    try {
        const result = resolveConfigImport({}, basedir, query);
        assert(typeof result === "string");
        resolved = result;
    } catch (error) {
        if (!isResolveError(error)) throw error;
        return { error: diagnosticFactories.couldNotResolveFrom(kind, query, basedir) };
    }

    const hasNoRequireHook = require.extensions[".ts"] === undefined;
    if (hasNoRequireHook && (resolved.endsWith(".ts") || resolved.endsWith(".tsx"))) {
        try {
            require("ts-node/register/transpile-only");
        } catch (err) {
            if (err.code !== "MODULE_NOT_FOUND") throw err;
            return { error: diagnosticFactories.toLoadItShouldBeTranspiled(kind, query) };
        }
    }

    const commonjsModule = require(resolved);
    const factoryModule = commonjsModule.__esModule ? commonjsModule : { default: commonjsModule };
    const result = factoryModule[importName];
    if (result === undefined) {
        return { error: diagnosticFactories.shouldHaveAExport(kind, query, importName) };
    }

    return { result };
}
