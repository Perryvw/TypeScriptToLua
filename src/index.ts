import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileWithSystem } from "./CommandLineParser";
import { CompilerOptions } from "./CompilerOptions";
import { getTranspileOutput, TranspilationResult, TranspiledFile } from "./Transpile";

export { parseCommandLine, ParsedCommandLine, updateParsedConfigFile } from "./CommandLineParser";
export { CompilerOptions, LuaLibImportKind, LuaTarget } from "./CompilerOptions";
export * from "./Emit";
export * from "./LuaAST";
export { LuaLibFeature } from "./LuaLib";
export { LuaPrinter } from "./LuaPrinter";
export { LuaTransformer } from "./LuaTransformer";
export * from "./Transpile";

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {}
): TranspilationResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { transpiledFiles, diagnostics: [...allDiagnostics] };
}

export function transpileProject(fileName: string, options?: CompilerOptions): TranspilationResult {
    const parseResult = parseConfigFileWithSystem(fileName, options);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, transpiledFiles: new Map() };
    }

    return transpileFiles(parseResult.fileNames, parseResult.options);
}

const libCache: { [key: string]: ts.SourceFile } = {};

/** @internal */
export function createVirtualProgram(
    input: Record<string, string>,
    options?: CompilerOptions
): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile: () => {},

        getSourceFile: filename => {
            if (filename in input) {
                return ts.createSourceFile(
                    filename,
                    input[filename],
                    ts.ScriptTarget.Latest,
                    false
                );
            }

            if (filename.startsWith("lib.")) {
                if (libCache[filename]) return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");

                libCache[filename] = ts.createSourceFile(
                    filename,
                    content,
                    ts.ScriptTarget.Latest,
                    false
                );

                return libCache[filename];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface VirtualProgramResult {
    file: TranspiledFile;
    diagnostics: ts.Diagnostic[];
}

export function transpileString(
    content: string,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const program = createVirtualProgram({ "main.ts": content }, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get("main.ts"), diagnostics: [...allDiagnostics] };
}

export function transpileVirtualProgram(
    files: Record<string, string>,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const mainFileName = Object.keys(files).find(x => /\bmain\.[a-z]+$/.test(x));
    if (mainFileName === undefined) {
        throw new Error('Virtual program should have a file named "main"');
    }

    const program = createVirtualProgram(files, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get(mainFileName), diagnostics: [...allDiagnostics] };
}
