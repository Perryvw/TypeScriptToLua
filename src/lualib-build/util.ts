import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
