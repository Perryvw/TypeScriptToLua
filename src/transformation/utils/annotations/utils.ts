import * as ts from "typescript";
import { AnnotationKind, getSignatureAnnotations, getTypeAnnotations } from ".";
import { TransformationContext } from "../../context";
import { findFirstNodeAbove, inferAssignedType } from "../typescript";

export function isTupleReturnCall(context: TransformationContext, node: ts.Node): boolean {
    if (!ts.isCallExpression(node)) {
        return false;
    }

    const signature = context.checker.getResolvedSignature(node);
    if (signature) {
        if (getSignatureAnnotations(context, signature).has(AnnotationKind.TupleReturn)) {
            return true;
        }

        // Only check function type for directive if it is declared as an interface or type alias
        const declaration = signature.getDeclaration();
        const isInterfaceOrAlias =
            declaration &&
            declaration.parent &&
            ((ts.isInterfaceDeclaration(declaration.parent) && ts.isCallSignatureDeclaration(declaration)) ||
                ts.isTypeAliasDeclaration(declaration.parent));
        if (!isInterfaceOrAlias) {
            return false;
        }
    }

    const type = context.checker.getTypeAtLocation(node.expression);
    return getTypeAnnotations(context, type).has(AnnotationKind.TupleReturn);
}

export function isInTupleReturnFunction(context: TransformationContext, node: ts.Node): boolean {
    const declaration = findFirstNodeAbove(node, ts.isFunctionLike);
    if (!declaration) {
        return false;
    }

    let functionType: ts.Type | undefined;
    if (ts.isFunctionExpression(declaration) || ts.isArrowFunction(declaration)) {
        functionType = inferAssignedType(context, declaration);
    } else if (ts.isMethodDeclaration(declaration) && ts.isObjectLiteralExpression(declaration.parent)) {
        // Manually lookup type for object literal properties declared with method syntax
        const interfaceType = inferAssignedType(context, declaration.parent);
        const propertySymbol = interfaceType.getProperty(declaration.name.getText());
        if (propertySymbol) {
            functionType = context.checker.getTypeOfSymbolAtLocation(propertySymbol, declaration);
        }
    }

    if (functionType === undefined) {
        functionType = context.checker.getTypeAtLocation(declaration);
    }

    // Check all overloads for directive
    const signatures = functionType.getCallSignatures();
    if (signatures && signatures.some(s => getSignatureAnnotations(context, s).has(AnnotationKind.TupleReturn))) {
        return true;
    }

    return getTypeAnnotations(context, functionType).has(AnnotationKind.TupleReturn);
}

export function isLuaIteratorType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(context, type).has(AnnotationKind.LuaIterator);
}

export function isVarArgType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(context, type).has(AnnotationKind.VarArg);
}

export function isForRangeType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(context, type).has(AnnotationKind.ForRange);
}
