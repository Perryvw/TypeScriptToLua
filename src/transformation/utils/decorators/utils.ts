import * as ts from "typescript";
import { TransformationContext } from "../../context";
import { DecoratorKind, getCustomDecorators, getCustomSignatureDecorators } from "../decorators";
import { findFirstNodeAbove, inferAssignedType } from "../typescript";

export function isTupleReturnCall(context: TransformationContext, node: ts.Node): boolean {
    if (!ts.isCallExpression(node)) {
        return false;
    }

    const signature = context.checker.getResolvedSignature(node);
    if (signature) {
        if (getCustomSignatureDecorators(context, signature).has(DecoratorKind.TupleReturn)) {
            return true;
        }

        // Only check function type for directive if it is declared as an interface or type alias
        const declaration = signature.getDeclaration();
        const isInterfaceOrAlias =
            declaration &&
            declaration.parent &&
            ((ts.isInterfaceDeclaration(declaration.parent) && ts.isCallSignatureDeclaration(declaration)) ||
                ts.isTypeAliasDeclaration(declaration.parent));
        if (!isInterfaceOrAlias) {
            return false;
        }
    }

    const type = context.checker.getTypeAtLocation(node.expression);
    return getCustomDecorators(context, type).has(DecoratorKind.TupleReturn);
}

export function isInTupleReturnFunction(context: TransformationContext, node: ts.Node): boolean {
    const declaration = findFirstNodeAbove(node, ts.isFunctionLike);
    if (!declaration) {
        return false;
    }

    let functionType: ts.Type | undefined;
    if (ts.isFunctionExpression(declaration) || ts.isArrowFunction(declaration)) {
        functionType = inferAssignedType(context, declaration);
    } else if (ts.isMethodDeclaration(declaration) && ts.isObjectLiteralExpression(declaration.parent)) {
        // Manually lookup type for object literal properties declared with method syntax
        const interfaceType = inferAssignedType(context, declaration.parent);
        const propertySymbol = interfaceType.getProperty(declaration.name.getText());
        if (propertySymbol) {
            functionType = context.checker.getTypeOfSymbolAtLocation(propertySymbol, declaration);
        }
    }

    if (functionType === undefined) {
        functionType = context.checker.getTypeAtLocation(declaration);
    }

    // Check all overloads for directive
    const signatures = functionType.getCallSignatures();
    if (signatures && signatures.some(s => getCustomSignatureDecorators(context, s).has(DecoratorKind.TupleReturn))) {
        return true;
    }

    const decorators = getCustomDecorators(context, functionType);
    return decorators.has(DecoratorKind.TupleReturn);
}

export function isLuaIteratorType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getCustomDecorators(context, type).has(DecoratorKind.LuaIterator);
}

export function isVarArgType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getCustomDecorators(context, type).has(DecoratorKind.Vararg);
}

export function isForRangeType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getCustomDecorators(context, type).has(DecoratorKind.ForRange);
}
