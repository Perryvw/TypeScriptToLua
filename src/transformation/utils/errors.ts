import * as ts from "typescript";
import { LuaTarget } from "../../CompilerOptions";

export class TranspileError extends Error {
    public name = "TranspileError";
    constructor(message: string, public node: ts.Node) {
        super(message);
    }
}

const getLuaTargetName = (version: LuaTarget) => (version === LuaTarget.LuaJIT ? "LuaJIT" : `Lua ${version}`);

export const InvalidDecoratorContext = (node: ts.Node) =>
    new TranspileError(`Decorator function cannot have 'this: void'.`, node);

export const MissingForOfVariables = (node: ts.Node) =>
    new TranspileError("Transpiled ForOf variable declaration list contains no declarations.", node);

export const UnsupportedForInVariable = (node: ts.Node) =>
    new TranspileError(`Unsupported for-in variable kind.`, node);

export const UndefinedScope = () => new Error("Expected to pop a scope, but found undefined.");

export const UnsupportedKind = (description: string, kind: ts.SyntaxKind, node: ts.Node) =>
    new TranspileError(`Unsupported ${description} kind: ${ts.SyntaxKind[kind]}`, node);

export const UnsupportedProperty = (parentName: string, property: string, node: ts.Node) =>
    new TranspileError(`Unsupported property on ${parentName}: ${property}`, node);

export const UnsupportedForTarget = (functionality: string, version: LuaTarget, node: ts.Node) =>
    new TranspileError(`${functionality} is/are not supported for target ${getLuaTargetName(version)}.`, node);

export const UnsupportedNoSelfFunctionConversion = (node: ts.Node, name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return new TranspileError(
        `Unable to convert function with a 'this' parameter to function${nameReference} with no 'this'. ` +
            `To fix, wrap in an arrow function, or declare with 'this: void'.`,
        node
    );
};

export const UnsupportedSelfFunctionConversion = (node: ts.Node, name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return new TranspileError(
        `Unable to convert function with no 'this' parameter to function${nameReference} with 'this'. ` +
            `To fix, wrap in an arrow function or declare with 'this: any'.`,
        node
    );
};

export const UnsupportedOverloadAssignment = (node: ts.Node, name?: string) => {
    const nameReference = name ? ` to '${name}'` : "";
    return new TranspileError(
        `Unsupported assignment of function with different overloaded types for 'this'${nameReference}. ` +
            "Overloads should all have the same type for 'this'.",
        node
    );
};

export const UnresolvableRequirePath = (node: ts.Node, reason: string, path?: string) =>
    new TranspileError(`${reason}. TypeScript path: ${path}.`, node);

export const ReferencedBeforeDeclaration = (node: ts.Identifier) =>
    new TranspileError(
        `Identifier "${node.text}" was referenced before it was declared. The declaration ` +
            "must be moved before the identifier's use, or hoisting must be enabled.",
        node
    );

export const UnsupportedObjectDestructuringInForOf = (node: ts.Node) =>
    new TranspileError(`Unsupported object destructuring in for...of statement.`, node);

export const InvalidAmbientIdentifierName = (node: ts.Identifier) =>
    new TranspileError(
        `Invalid ambient identifier name "${node.text}". Ambient identifiers must be valid lua identifiers.`,
        node
    );
