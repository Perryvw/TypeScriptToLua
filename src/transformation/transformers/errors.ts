import * as ts from "typescript";
import * as tstl from "../../LuaAST";
import { FunctionVisitor, TransformerPlugin } from "../context";
import { isInTupleReturnFunction } from "../utils/annotations";
import { InvalidThrowExpression } from "../utils/errors";
import { createUnpackCall } from "../utils/lua-ast";
import { findScope, ScopeType } from "../utils/scope";
import { isStringType } from "../utils/typescript";
import { transformScopeBlock } from "./block";
import { transformIdentifier } from "./identifier";

const transformTryStatement: FunctionVisitor<ts.TryStatement> = (statement, context) => {
    const [tryBlock, tryScope] = transformScopeBlock(context, statement.tryBlock, ScopeType.Try);

    const tryResultIdentifier = tstl.createIdentifier("____try");
    const returnValueIdentifier = tstl.createIdentifier("____returnValue");

    const result: tstl.Statement[] = [];

    let returnedIdentifier: tstl.Identifier | undefined;
    let returnCondition: tstl.Expression | undefined;

    const pCall = tstl.createIdentifier("pcall");
    const tryCall = tstl.createCallExpression(pCall, [tstl.createFunctionExpression(tryBlock)]);

    if (statement.catchClause && statement.catchClause.block.statements.length > 0) {
        // try with catch
        let [catchBlock, catchScope] = transformScopeBlock(context, statement.catchClause.block, ScopeType.Catch);
        if (statement.catchClause.variableDeclaration) {
            // Replace ____returned with catch variable
            returnedIdentifier = transformIdentifier(context, statement.catchClause.variableDeclaration
                .name as ts.Identifier);
        } else if (tryScope.functionReturned || catchScope.functionReturned) {
            returnedIdentifier = tstl.createIdentifier("____returned");
        }

        const tryReturnIdentifiers = [tryResultIdentifier]; // ____try
        if (returnedIdentifier) {
            tryReturnIdentifiers.push(returnedIdentifier); // ____returned or catch variable
            if (tryScope.functionReturned || catchScope.functionReturned) {
                tryReturnIdentifiers.push(returnValueIdentifier); // ____returnValue
                returnCondition = tstl.cloneIdentifier(returnedIdentifier);
            }
        }
        result.push(tstl.createVariableDeclarationStatement(tryReturnIdentifiers, tryCall));

        if ((tryScope.functionReturned || catchScope.functionReturned) && returnedIdentifier) {
            // Wrap catch in function if try or catch has return
            const catchCall = tstl.createCallExpression(
                tstl.createParenthesizedExpression(tstl.createFunctionExpression(catchBlock))
            );
            const catchAssign = tstl.createAssignmentStatement(
                [tstl.cloneIdentifier(returnedIdentifier), tstl.cloneIdentifier(returnValueIdentifier)],
                catchCall
            );
            catchBlock = tstl.createBlock([catchAssign]);
        }
        const notTryCondition = tstl.createUnaryExpression(
            tstl.createParenthesizedExpression(tryResultIdentifier),
            tstl.SyntaxKind.NotOperator
        );
        result.push(tstl.createIfStatement(notTryCondition, catchBlock));
    } else if (tryScope.functionReturned) {
        // try with return, but no catch
        returnedIdentifier = tstl.createIdentifier("____returned");
        const returnedVariables = [tryResultIdentifier, returnedIdentifier, returnValueIdentifier];
        result.push(tstl.createVariableDeclarationStatement(returnedVariables, tryCall));

        // change return condition from '____returned' to '____try and ____returned'
        returnCondition = tstl.createBinaryExpression(
            tstl.cloneIdentifier(tryResultIdentifier),
            returnedIdentifier,
            tstl.SyntaxKind.AndOperator
        );
    } else {
        // try without return or catch
        result.push(tstl.createExpressionStatement(tryCall));
    }

    if (statement.finallyBlock && statement.finallyBlock.statements.length > 0) {
        result.push(...context.transformStatements(statement.finallyBlock));
    }

    if (returnCondition && returnedIdentifier) {
        // With catch clause:
        //     if ____returned then return ____returnValue end
        // No catch clause:
        //     if ____try and ____returned then return ____returnValue end
        const returnValues: tstl.Expression[] = [];
        const parentTryCatch = findScope(context, ScopeType.Function | ScopeType.Try | ScopeType.Catch);
        if (parentTryCatch && parentTryCatch.type !== ScopeType.Function) {
            // Nested try/catch needs to prefix a 'true' return value
            returnValues.push(tstl.createBooleanLiteral(true));
        }

        if (isInTupleReturnFunction(context, statement)) {
            returnValues.push(createUnpackCall(context, tstl.cloneIdentifier(returnValueIdentifier)));
        } else {
            returnValues.push(tstl.cloneIdentifier(returnValueIdentifier));
        }

        const returnStatement = tstl.createReturnStatement(returnValues);
        const ifReturnedStatement = tstl.createIfStatement(returnCondition, tstl.createBlock([returnStatement]));
        result.push(ifReturnedStatement);
    }

    return tstl.createDoStatement(result, statement);
};

const transformThrowStatement: FunctionVisitor<ts.ThrowStatement> = (statement, context) => {
    if (statement.expression === undefined) {
        throw InvalidThrowExpression(statement);
    }

    const type = context.checker.getTypeAtLocation(statement.expression);
    if (!isStringType(context, type)) {
        throw InvalidThrowExpression(statement.expression);
    }

    const error = tstl.createIdentifier("error");
    return tstl.createExpressionStatement(
        tstl.createCallExpression(error, [context.transformExpression(statement.expression)]),
        statement
    );
};

export const errorsPlugin: TransformerPlugin = {
    visitors: {
        [ts.SyntaxKind.TryStatement]: transformTryStatement,
        [ts.SyntaxKind.ThrowStatement]: transformThrowStatement,
    },
};
