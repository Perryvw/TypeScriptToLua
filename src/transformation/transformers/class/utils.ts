import * as ts from "typescript";
import { DecoratorKind, getCustomDecorators } from "../../utils/decorators";
import { TransformationContext } from "../../context";

export function isStaticNode(node: ts.Node): boolean {
    return node.modifiers !== undefined && node.modifiers.some(m => m.kind === ts.SyntaxKind.StaticKeyword);
}

export function getExtendedTypeNode(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.ExpressionWithTypeArguments | undefined {
    if (node && node.heritageClauses) {
        for (const clause of node.heritageClauses) {
            if (clause.token === ts.SyntaxKind.ExtendsKeyword) {
                const superType = context.checker.getTypeAtLocation(clause.types[0]);
                const decorators = getCustomDecorators(context, superType);
                if (!decorators.has(DecoratorKind.PureAbstract)) {
                    return clause.types[0];
                }
            }
        }
    }
}

export function getExtendedType(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.Type | undefined {
    const extendedTypeNode = getExtendedTypeNode(context, node);
    return extendedTypeNode && context.checker.getTypeAtLocation(extendedTypeNode);
}

export function findInClassOrAncestor(
    context: TransformationContext,
    classDeclaration: ts.ClassLikeDeclarationBase,
    callback: (classDeclaration: ts.ClassLikeDeclarationBase) => boolean
): ts.ClassLikeDeclarationBase | undefined {
    if (callback(classDeclaration)) {
        return classDeclaration;
    }

    const extendsType = getExtendedType(context, classDeclaration);
    if (!extendsType) {
        return undefined;
    }

    const symbol = extendsType.getSymbol();
    if (symbol === undefined) {
        return undefined;
    }

    const symbolDeclarations = symbol.getDeclarations();
    if (symbolDeclarations === undefined) {
        return undefined;
    }

    const declaration = symbolDeclarations.find(ts.isClassLike);
    if (!declaration) {
        return undefined;
    }

    return findInClassOrAncestor(context, declaration, callback);
}
