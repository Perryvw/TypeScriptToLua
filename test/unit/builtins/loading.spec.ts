import * as tstl from "../../../src";
import { UnsupportedProperty } from "../../../src/transformation/utils/errors";
import * as util from "../../util";

describe("luaLibImport", () => {
    test("inline", () => {
        util.testExpression`[0].push(1)`
            .setOptions({ luaLibImport: tstl.LuaLibImportKind.Inline })
            .tap(builder => expect(builder.getMainLuaCodeChunk()).not.toContain(`require("lualib_bundle")`))
            .expectToMatchJsResult();
    });

    test("require", () => {
        util.testExpression`[0].push(1)`
            .setOptions({ luaLibImport: tstl.LuaLibImportKind.Require })
            .tap(builder => expect(builder.getMainLuaCodeChunk()).toContain(`require("lualib_bundle")`))
            .expectToMatchJsResult();
    });

    test("always", () => {
        util.testModule``
            .setOptions({ luaLibImport: tstl.LuaLibImportKind.Always })
            .tap(builder => expect(builder.getMainLuaCodeChunk()).toContain(`require("lualib_bundle")`))
            .expectToEqual(undefined);
    });
});

test.each([tstl.LuaLibImportKind.Inline, tstl.LuaLibImportKind.None, tstl.LuaLibImportKind.Require])(
    "should not include lualib without code (%p)",
    luaLibImport => {
        util.testModule``.setOptions({ luaLibImport }).tap(builder => expect(builder.getMainLuaCodeChunk()).toBe(""));
    }
);

test("lualib should not include tstl header", () => {
    util.testExpression`[0].push(1)`.tap(builder =>
        expect(builder.getMainLuaCodeChunk()).not.toContain("Generated with")
    );
});

describe("Unknown builtin property", () => {
    test("access", () => {
        util.testExpression`Math.unknownProperty`
            .disableSemanticCheck()
            .expectToHaveDiagnosticOfError(UnsupportedProperty("Math", "unknownProperty", util.nodeStub));
    });
});
