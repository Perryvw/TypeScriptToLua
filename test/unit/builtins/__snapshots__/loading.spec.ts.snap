// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unknown builtin property access: code 1`] = `
"local ____exports = {}
____exports.__result = Math.unknownProperty
return ____exports"
`;

exports[`Unknown builtin property access: diagnostics 1`] = `"main.ts(1,25): error TSTL: Math.unknownProperty is unsupported."`;

exports[`Unknown builtin property function call: code 1`] = `
"local ____exports = {}
____exports.__result = ({}):unknownFunction()
return ____exports"
`;

exports[`Unknown builtin property function call: diagnostics 1`] = `"main.ts(1,25): error TSTL: array.unknownFunction is unsupported."`;
